src/mscorlib/shared/Internal/Runtime/CompilerServices/Unsafe.cs

```diff
@@ -94,117 +100,230 @@ public static int SizeOf<T>()
+            // See getILIntrinsicImplementationForUnsafe for how this happens.
             typeof(T).ToString(); // Type token used by the actual method body
-            throw new InvalidOperationException();
+            throw new PlatformNotSupportedException();
+#else
+            return Unsafe.As<byte, T>(ref source);
+#endif
         }
 
-        [CLSCompliant(false)]
+        /// <summary>
+        /// Writes a value of type <typeparamref name="T"/> to the given location.
+        /// </summary>
+        [Intrinsic]
         [NonVersionable]
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static void WriteUnaligned<T>(void* destination, T value)
         {
+#if CORECLR
             // The body of this function will be replaced by the EE with unsafe code!!!
-            // See getILIntrinsicImplementationForUnsafe for how this happens.  
+            // See getILIntrinsicImplementationForUnsafe for how this happens.
             typeof(T).ToString(); // Type token used by the actual method body
-            throw new InvalidOperationException();
+            throw new PlatformNotSupportedException();
+#else
+            Unsafe.As<byte, T>(ref *(byte*)destination) = value;
+#endif
         }
 
+        /// <summary>
+        /// Writes a value of type <typeparamref name="T"/> to the given location.
+        /// </summary>
+        [Intrinsic]
         [NonVersionable]
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static void WriteUnaligned<T>(ref byte destination, T value)
         {
+#if CORECLR
             // The body of this function will be replaced by the EE with unsafe code!!!
-            // See getILIntrinsicImplementationForUnsafe for how this happens.  
+            // See getILIntrinsicImplementationForUnsafe for how this happens.
             typeof(T).ToString(); // Type token used by the actual method body
-            throw new InvalidOperationException();
+            throw new PlatformNotSupportedException();
+#else
+            Unsafe.As<byte, T>(ref destination) = value;
+#endif
+        }
+
+        /// <summary>
+        /// Adds an element offset to the given reference.
+        /// </summary>
+        [Intrinsic]
+        [NonVersionable]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static ref T AddByteOffset<T>(ref T source, IntPtr byteOffset)
+        {
+            // This method is implemented by the toolchain
+            throw new PlatformNotSupportedException();
+
+            // ldarg.0
+            // ldarg.1
+            // add
+            // ret
+        }
+
+        /// <summary>
+        /// Reads a value of type <typeparamref name="T"/> from the given location.
+        /// </summary>
+        [Intrinsic]
+        [NonVersionable]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static T Read<T>(void* source)
```

Comment:
These can be public. (I think that actually have to be public - it would cause a build break in ProjectN otherwise.)

Reply:
And the implementation can remain common between CoreCLR/ProjectN?

Reply:
Yes. (Note that it does not prevent the runtime to provide more streamlined implementation.)